src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1427,126 +1433,407 @@ let CheckMultipleInputsSequential (ctok, checkForErrors, tcConfig, tcImports, tc
+                    else
+                        mkPhysicalFile dep)
+                deps
+
+        // Transform the graph to include ArtificialImplFile nodes when necessary.
+        graph
+        |> Seq.collect (fun (KeyValue (fileIdx, deps)) ->
+            if filePairs.IsSignature fileIdx then
+                // Add an additional ArtificialImplFile node for the signature file.
+                [|
+                    // Mark the current file as physical and map the dependencies.
+                    mkPhysicalFile fileIdx, mapDependencies fileIdx deps
+                    // Introduce a new node that depends on the signature.
+                    mkArtificialImplFile fileIdx, [| mkPhysicalFile fileIdx |]
+                |]
+            else
+                [| mkPhysicalFile fileIdx, mapDependencies fileIdx deps |])
+        |> Graph.make
+
+    // Persist the graph to a Mermaid diagram if specified.
+    if tcConfig.typeCheckingConfig.DumpGraph then
+        tcConfig.outputFile
+        |> Option.iter (fun outputFile ->
+            let outputFile = FileSystem.GetFullPathShim(outputFile)
+            let graphFile = FileSystem.ChangeExtensionShim(outputFile, ".graph.md")
+
+            graph
+            |> Graph.map (fun idx ->
+                let friendlyFileName =
+                    sourceFiles[idx]
+                        .FileName.Replace(tcConfig.implicitIncludeDir, "")
+                        .TrimStart([| '\\'; '/' |])
+
+                (idx, friendlyFileName))
+            |> Graph.serialiseToMermaid graphFile)
+
+    let _ = ctok // TODO Use it
+    let diagnosticsLogger = DiagnosticsThreadStatics.DiagnosticsLogger
+
+    // In the first linear part of parallel checking, we use a 'checkForErrors' that checks either for errors
+    // somewhere in the files processed prior to each one, or in the processing of this particular file.
+    let priorErrors = checkForErrors ()
+
+    let mutable cnt = 1
+
+    let processArtificialImplFile (input: ParsedInput) ((currentTcState, _currentPriorErrors): State) : Finisher<State, PartialResult> =
+        Finisher(fun (state: State) ->
+            let tcState, currentPriorErrors = state
+
+            let f =
+                // Retrieve the type-checked signature information and add it to the TcEnvFromImpls.
+                AddSignatureResultToTcImplEnv(tcImports, tcGlobals, prefixPathOpt, TcResultsSink.NoSink, currentTcState, input)
+
+            // The `partialResult` will be excluded at the end of `GraphProcessing.processGraph`.
+            // The important thing is that `nextTcState` will populated the necessary information to TcEnvFromImpls.
+            let partialResult, nextTcState = f tcState
+            partialResult, (nextTcState, currentPriorErrors))
+
+    let processFile
+        ((input, logger): ParsedInput * DiagnosticsLogger)
+        ((currentTcState, _currentPriorErrors): State)
+        : Finisher<State, PartialResult> =
+        cancellable {
```

Comment:
Maybe this whole block doesn't have to be wrapped in `cancellable` but we can just call `Cancellable.runWithoutCancellation` on `CheckOneInputWithCallback`?